doc/story008-repl-traces.md

```diff
@@ -0,0 +1,262 @@
+20
+```
+
+**Shortcut.** Since typing `q::state` often may be tedious, we offer a shortcut
+*`q::s` for `q::state`.
+
+## Story 2: Inspect the trace
+
+*As a protocol reader, I want to print the sequence of states that were visited
+on the way to the current protocol state, so I can see how the system was
+progressing.*
+
+Assume that we have executed a few steps:
+
+```bluespec
+>>> init
+true
+>>> pressButton
+true
+>>> heat
+true
+>>> depressButton
+true
+```
+
+**Current interface**. After executing a few steps, we can inspect the reached
+state. However, we would not be able to inspect the intermediate states, e.g.,
+the states reached by executing `init`, then `pressButton`, then `heat`.
+
+As mentioned in [#446][], it would be great to observe the sequence of visited
+states. Currently, the only way to do that is by replaying the actions from the
+start and printing the intermediate states after executing every action.
+
+**Proposed interface**. Similar to Story 1, we are proposing to use a standard
+name such as `q::trace`.
+
+For the sequence `init.then(pressButton).then(heat).then(depressButton)`, we
+expect `q::trace` to evaluate as follows:
+
+```bluespec
+>>> q::trace
+[
+  // 0
+  { beeping: false, heatingOn: false, temperature: 20 },
+  // 1
+  { beeping: false, heatingOn: true, temperature: 20 },
+  // 2
+  { beeping: false, heatingOn: true, temperature: 21 },
+  // 3*
+  { beeping: false, heatingOn: false, temperature: 21 },
+]
+```
+
+As we can see, `q::trace` is simply a value, which is printed as a
+syntactically correct Quint expression. We expect REPL to decorate it a bit with
+helpful comments: It should print the state numbers and highlight the active
+state (with a `*`). However, the fact that `q::trace` is a value, lets us to
+access it programmatically:
+
+```bluespec
+>>> q::trace.length()
+4
+>>> q::trace[3].temperature - q::trace[0].temperature
+1
+```
+
+**Shortcut.** Since typing `q::trace` may be tedious, we offer a shortcut
+`q::t` for `q::trace`.
+
+## Story 3: Navigate the trace
+
+*As a protocol reader, I want to jump between the visited states in the trace,
+so I can evaluate invariants against them and explore different paths.*
+
+As in Story 2, assume that we have executed a few steps:
+
+```bluespec
+>>> init
+true
+>>> pressButton
+true
+>>> heat
+true
+>>> depressButton
+true
+```
+
+**Current interface**. Assuming that `q::trace` from Story 2 is implemented,
+we should be able to see the intermediate states that are produced by the above
+sequence of actions. However, if we want to evaluate state invariants against
+the intermediate states, we have to replay the actions from start, e.g., by
+executing `init`, then `pressButton`, then `heat`.
+
+**Proposed interface**. We introduce several actions for navigating the trace.
+
+### 3.1. Jump
+
+The following REPL session highlights the expected interface:
+
+```bluespec
+>>> init.then(pressButton).then(heat).then(depressButton)
+true
+>>> q::jump(2)
```

Comment:
While I agree with @thpani that these features may be confusing for newcomers, I would love these features. I'd volte for implementing and clearly document it rather than not.

On the interest of clearly documenting, I would do `::jump(1)` here so that when the history is modified below the history is shorter, showing very clearly that something was thrown away.